### ISSUE TITLE
Recurrent uncaught exception in choose_metric

### ISSUE THREAD

**Issue by benjolitz** [graphite-project/carbon#815]
Found this spamming my syslog:

```
Oct  6 11:40:24 nyx carbon-cache-a[1193]: [console] Unhandled Error
Oct  6 11:40:24 nyx carbon-cache-a[1193]: [console] 	Traceback (most recent call last):
Oct  6 11:40:24 nyx carbon-cache-a[1193]: [console] 	  File "/usr/local/lib/python3.5/site-packages/twisted/python/threadpool.py", line 250, in inContext
Oct  6 11:40:24 nyx carbon-cache-a[1193]: [console] 	    result = inContext.theWork()
Oct  6 11:40:24 nyx carbon-cache-a[1193]: [console] 	  File "/usr/local/lib/python3.5/site-packages/twisted/python/threadpool.py", line 266, in <lambda>
Oct  6 11:40:24 nyx carbon-cache-a[1193]: [console] 	    inContext.theWork = lambda: context.call(ctx, func, *args, **kw)
Oct  6 11:40:24 nyx carbon-cache-a[1193]: [console] 	  File "/usr/local/lib/python3.5/site-packages/twisted/python/context.py", line 122, in callWithContext
Oct  6 11:40:24 nyx carbon-cache-a[1193]: [console] 	    return self.currentContext().callWithContext(ctx, func, *args, **kw)
Oct  6 11:40:24 nyx carbon-cache-a[1193]: [console] 	  File "/usr/local/lib/python3.5/site-packages/twisted/python/context.py", line 85, in callWithContext
Oct  6 11:40:24 nyx carbon-cache-a[1193]: [console] 	    return func(*args,**kw)
Oct  6 11:40:24 nyx carbon-cache-a[1193]: [console] 	--- <exception caught here> ---
Oct  6 11:40:24 nyx carbon-cache-a[1193]: [console] 	  File "/opt/graphite/lib/carbon/writer.py", line 188, in writeForever
Oct  6 11:40:24 nyx carbon-cache-a[1193]: [console] 	    writeCachedDataPoints()
Oct  6 11:40:24 nyx carbon-cache-a[1193]: [console] 	  File "/opt/graphite/lib/carbon/writer.py", line 97, in writeCachedDataPoints
Oct  6 11:40:24 nyx carbon-cache-a[1193]: [console] 	    (metric, datapoints) = cache.drain_metric()
Oct  6 11:40:24 nyx carbon-cache-a[1193]: [console] 	  File "/opt/graphite/lib/carbon/cache.py", line 187, in drain_metric
Oct  6 11:40:24 nyx carbon-cache-a[1193]: [console] 	    metric = self.strategy.choose_item()
Oct  6 11:40:24 nyx carbon-cache-a[1193]: [console] 	  File "/opt/graphite/lib/carbon/cache.py", line 116, in choose_item
Oct  6 11:40:24 nyx carbon-cache-a[1193]: [console] 	    return next(self.queue)
Oct  6 11:40:24 nyx carbon-cache-a[1193]: [console] 	builtins.StopIteration:
```

Despite restarting the carbon-cache daemon (`/opt/graphite/bin/carbon-cache.py --nodaemon --syslog start`), the issue reappeared an hour later.

Switching to `CACHE_WRITE_STRATEGY = naive` from `sorted` seemingly dealt with the issue.

This is on a low load box (60 metrics / second).

**Comment by piotr1212** [comment]
Which version are you using?

**Comment by deniszh** [comment]
Thanks! Quite fresh, though. Any ideas, @piotr1212 ?

**Comment by piotr1212** [comment]
I don't see how this can be triggered without the queue being empty. @benjolitz could you please share your carbon.conf?

**Comment by benjolitz** [comment]
Sure:

```
[cache]
STORAGE_DIR = /Archive2/volatile/storage/
DATABASE = whisper
ENABLE_LOGROTATION = True
USER = www
GRAPHITE_URL = http://localhost:80/graphite/
MAX_CACHE_SIZE = inf
MAX_UPDATES_PER_SECOND = 500
MAX_CREATES_PER_MINUTE = 50
LINE_RECEIVER_INTERFACE = 0.0.0.0
LINE_RECEIVER_PORT = 2003
ENABLE_UDP_LISTENER = False
UDP_RECEIVER_INTERFACE = 0.0.0.0
UDP_RECEIVER_PORT = 2003
PICKLE_RECEIVER_INTERFACE = 0.0.0.0
PICKLE_RECEIVER_PORT = 2004
USE_INSECURE_UNPICKLER = False
CACHE_QUERY_INTERFACE = 0.0.0.0
CACHE_QUERY_PORT = 7002
USE_FLOW_CONTROL = True
LOG_UPDATES = False
LOG_CACHE_HITS = False
LOG_CACHE_QUEUE_SORTS = False
CACHE_WRITE_STRATEGY = naive
WHISPER_AUTOFLUSH = False
WHISPER_FALLOCATE_CREATE = False

[relay]
LINE_RECEIVER_INTERFACE = 0.0.0.0
LINE_RECEIVER_PORT = 2013
PICKLE_RECEIVER_INTERFACE = 0.0.0.0
PICKLE_RECEIVER_PORT = 2014
RELAY_METHOD = rules
REPLICATION_FACTOR = 1
DESTINATIONS = 127.0.0.1:2004
MAX_QUEUE_SIZE = 10000
MAX_DATAPOINTS_PER_MESSAGE = 500
QUEUE_LOW_WATERMARK_PCT = 0.8
TIME_TO_DEFER_SENDING = 0.0001
USE_FLOW_CONTROL = True
USE_RATIO_RESET=False
MIN_RESET_STAT_FLOW=1000
MIN_RESET_RATIO=0.9
MIN_RESET_INTERVAL=121

[aggregator]
LINE_RECEIVER_INTERFACE = 0.0.0.0
LINE_RECEIVER_PORT = 2023
PICKLE_RECEIVER_INTERFACE = 0.0.0.0
PICKLE_RECEIVER_PORT = 2024
FORWARD_ALL = True
DESTINATIONS = 127.0.0.1:2004
REPLICATION_FACTOR = 1
MAX_QUEUE_SIZE = 10000
USE_FLOW_CONTROL = True
MAX_DATAPOINTS_PER_MESSAGE = 500
MAX_AGGREGATION_INTERVALS = 5
```

**Comment by piotr1212** [comment]
So this can happen when an exception gets raised in the generator here https://github.com/graphite-project/carbon/blob/master/lib/carbon/cache.py#L101-L111 . That breaks the generator and raises the StopIterationException. I haven't figured out what can cause the exception but expect that the queue has some malformed data which fails on either the sort or the itemgetter call, not sure yet.

**Comment by piotr1212** [comment]
Any chance if you could run following patch and check your logs then? That would be really helpful.

 https://github.com/piotr1212/carbon/commit/bf64fea0a5be107fa67b336d53b1580354e36c2a

**Comment by benjolitz** [comment]
I've added the try/except guards around the function cited in the above commit along with the log.err() call. Also switched back to sorted strategy.

**Comment by piotr1212** [comment]
Thanks, please let me know what you'll find in your logs.

**Comment by piotr1212** [comment]
Did you get any new stack traces in your logs?

**Comment by ma-tty** [comment]
The issue is related to #822. The first "builtins.RuntimeError: dictionary changed size during iteration" error  appears in logs only once followed by repeated "builtins.StopIteration:" errors. So it could be hard to see the first error:
```
11/11/2018 00:01:45 :: [console] Unhandled Error
Traceback (most recent call last):
  File "/opt/graphite/lib/python3.6/site-packages/twisted/python/threadpool.py", line 250, in inContext
    result = inContext.theWork()
  File "/opt/graphite/lib/python3.6/site-packages/twisted/python/threadpool.py", line 266, in <lambda>
    inContext.theWork = lambda: context.call(ctx, func, *args, **kw)
  File "/opt/graphite/lib/python3.6/site-packages/twisted/python/context.py", line 122, in callWithContext
    return self.currentContext().callWithContext(ctx, func, *args, **kw)
  File "/opt/graphite/lib/python3.6/site-packages/twisted/python/context.py", line 85, in callWithContext
    return func(*args,**kw)
--- <exception caught here> ---
  File "/opt/graphite/lib/carbon/writer.py", line 189, in writeForever
    writeCachedDataPoints()
  File "/opt/graphite/lib/carbon/writer.py", line 98, in writeCachedDataPoints
    (metric, datapoints) = cache.drain_metric()
  File "/opt/graphite/lib/carbon/cache.py", line 187, in drain_metric
    metric = self.strategy.choose_item()
  File "/opt/graphite/lib/carbon/cache.py", line 116, in choose_item
    return next(self.queue)
  File "/opt/graphite/lib/carbon/cache.py", line 104, in _generate_queue
    metric_counts = sorted(self.cache.counts, key=lambda x: x[1])
  File "/opt/graphite/lib/carbon/cache.py", line 161, in counts
    return [(metric, len(datapoints)) for (metric, datapoints) in self.items()]
  File "/opt/graphite/lib/carbon/cache.py", line 161, in <listcomp>
    return [(metric, len(datapoints)) for (metric, datapoints) in self.items()]
builtins.RuntimeError: dictionary changed size during iteration


11/11/2018 00:01:45 :: [console] Unhandled Error
Traceback (most recent call last):
  File "/opt/graphite/lib/python3.6/site-packages/twisted/python/threadpool.py", line 250, in inContext
    result = inContext.theWork()
  File "/opt/graphite/lib/python3.6/site-packages/twisted/python/threadpool.py", line 266, in <lambda>
    inContext.theWork = lambda: context.call(ctx, func, *args, **kw)
  File "/opt/graphite/lib/python3.6/site-packages/twisted/python/context.py", line 122, in callWithContext
    return self.currentContext().callWithContext(ctx, func, *args, **kw)
  File "/opt/graphite/lib/python3.6/site-packages/twisted/python/context.py", line 85, in callWithContext
    return func(*args,**kw)
--- <exception caught here> ---
  File "/opt/graphite/lib/carbon/writer.py", line 189, in writeForever
    writeCachedDataPoints()
  File "/opt/graphite/lib/carbon/writer.py", line 98, in writeCachedDataPoints
    (metric, datapoints) = cache.drain_metric()
  File "/opt/graphite/lib/carbon/cache.py", line 187, in drain_metric
    metric = self.strategy.choose_item()
  File "/opt/graphite/lib/carbon/cache.py", line 116, in choose_item
    return next(self.queue)
builtins.StopIteration:
    
11/11/2018 00:01:45 :: [console] Unhandled Error
Traceback (most recent call last):
  File "/opt/graphite/lib/python3.6/site-packages/twisted/python/threadpool.py", line 250, in inContext
    result = inContext.theWork()
  File "/opt/graphite/lib/python3.6/site-packages/twisted/python/threadpool.py", line 266, in <lambda>
    inContext.theWork = lambda: context.call(ctx, func, *args, **kw)
  File "/opt/graphite/lib/python3.6/site-packages/twisted/python/context.py", line 122, in callWithContext
    return self.currentContext().callWithContext(ctx, func, *args, **kw)
  File "/opt/graphite/lib/python3.6/site-packages/twisted/python/context.py", line 85, in callWithContext
    return func(*args,**kw)
--- <exception caught here> ---
  File "/opt/graphite/lib/carbon/writer.py", line 189, in writeForever
    writeCachedDataPoints()
  File "/opt/graphite/lib/carbon/writer.py", line 98, in writeCachedDataPoints
    (metric, datapoints) = cache.drain_metric()
  File "/opt/graphite/lib/carbon/cache.py", line 187, in drain_metric
    metric = self.strategy.choose_item()
  File "/opt/graphite/lib/carbon/cache.py", line 116, in choose_item
    return next(self.queue)
builtins.StopIteration:
```

**Comment by deniszh** [comment]
Fix released in 1.1.5
